### PR TITLE
Feature/drpc

### PR DESCRIPTION
**Pull request by mattdean-digicatapult**
# Pull Request

## Checklist
- [x] Have you read Digital Catapult's [Code of Conduct](https://github.com/digicatapult/.github/blob/main/CODE_OF_CONDUCT.md)?
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] New and existing unit tests pass locally with my changes.

## PR Type

Please delete options that are irrelevant.

- [ ] Bug Fix
- [x] Feature
- [ ] Documentation Update
- [ ] Code style update (formatting, local variables)
- [ ] Breaking Change (fix or feature that would cause existing functionality to change)

## Linked tickets

[VR-176](https://digicatapult.atlassian.net/browse/VR-176)

## High level description

Drpc endpoints to suport sending veritable queries

## Detailed description

Adds Drpc endpoints for sending and repsonding to drpc requests

## Describe alternatives you've considered

Considered avoiding the new endpoints by re-using the verified-drpc functionality with a new handler implementation. That would however have involved working out how to generify the proof expectations

## Operational impact

None

## Additional context

N/A

[VR-176]: https://digicatapult.atlassian.net/browse/VR-176?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/controllers/v1/drpc/DrpcController.ts [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+   * @param timeout The timeout for receiving a response
+   */
+  @Post('/:connectionId/request')
+  @Response<NotFound>(404)
+  @Response<BadGatewayError>(502)
+  @Response<GatewayTimeout>(504)
+  public async sendRequest(
+    @Path('connectionId') connectionId: RecordId,
+    @Body() requestOptions: DrpcRequestOptions,
+    @Query('timeout') timeout = 5000
```
Comment (rmlearney-digicatapult) [inline src/controllers/v1/drpc/DrpcController.ts:56]: Do we need a timeout/synchrony or do we stay true to more pure 'offline-first' DIDComm?

Comment (mattdean-digicatapult) [reply src/controllers/v1/drpc/DrpcController.ts:56]: The way the drpc API in credo works this is almost impossible. There's no way of listing drpc requests and no way of getting any information about the request except via the returned `responseListener` function. As such we must assume a short timeout between request and response. Offline first is explicitly prevented by the API design

Comment (rmlearney-digicatapult) [reply src/controllers/v1/drpc/DrpcController.ts:56]: Hmm. I wonder if they discussed this during the development of their implementation. Will try to find a thread. Perhaps they were working within a limitation of DIDComm v1?

File: src/drpc-handler/index.ts [hunk 2]
```diff
@@ -1,5 +1,85 @@
...
+type ReceivedRequest = {
+  request: DrpcRequestObject
+  sendResponse: (response: DrpcResponse) => Promise<void>
+}
+
+@injectable()
+@singleton()
+export default class DrpcReceiveHandler {
+  private agent: RestAgent
+  private requests = new Map<string | number, ReceivedRequest>()
```
[the remark this answers is not in the document]
Comment (mattdean-digicatapult) [reply src/drpc-handler/index.ts:23]: there's a function called `loop` below that's calling `this.agent.modules.drpc.recvRequest()` in a loop to receive requests beform some crude validation and then add them to the map. When a response is triggered we remove items from the map

File: src/controllers/v1/drpc/DrpcController.ts (hunk 1 above)
Comment (rmlearney-digicatapult) [reply src/controllers/v1/drpc/DrpcController.ts:56]: According to the spec at https://github.com/hyperledger/aries-rfcs/blob/main/features/0804-didcomm-rpc/README.md, there is a required @type decorator which must equal `"@type": "https://didcomm.org/drpc/1.0/request"
`

Comment (mattdean-digicatapult) [reply src/controllers/v1/drpc/DrpcController.ts:56]: Yes but that's not the credo-ts api that's what's happening under the hood. Look at how `recvRequest` works https://github.com/openwallet-foundation/credo-ts/tree/main/packages/drpc.

**Review by Ellenn-A — APPROVED** [review]
lgtm

**Pull request merged**